vespa/src/main/java/ai/marqo/search/HybridSearcher.java

```diff
@@ -0,0 +1,457 @@
+            }
+
+        } else if (STANDARD_SEARCH_TYPES.contains(retrievalMethod)) {
+            if (STANDARD_SEARCH_TYPES.contains(rankingMethod)) {
+                Query combinedQuery =
+                        createSubQuery(query, retrievalMethod, rankingMethod, verbose);
+                Result result = execution.search(combinedQuery);
+                logIfVerbose("Results: ", verbose);
+                logHitGroup(result.hits(), verbose);
+                return execution.search(combinedQuery);
+            } else {
+                throw new RuntimeException(
+                        "If retrievalMethod is 'lexical' or 'tensor', rankingMethod can only be"
+                                + " 'lexical', or 'tensor'.");
+            }
+        } else {
+            throw new RuntimeException(
+                    "retrievalMethod can only be 'disjunction', 'lexical', or 'tensor'.");
+        }
+    }
+
+    /**
+     * Implement feature score scaling and normalization
+     * @param hitsTensor
+     * @param hitsLexical
+     * @param k
+     * @param alpha
+     * @param verbose
+     */
+    HitGroup rrf(
+            HitGroup hitsTensor, HitGroup hitsLexical, Integer k, Double alpha, boolean verbose) {
+
+        HashMap<String, Double> rrfScores = new HashMap<>();
+        HitGroup result = new HitGroup();
+        Double reciprocalRank, existingScore, newScore;
+
+        logIfVerbose("Beginning RRF process.", verbose);
+        logIfVerbose("Beginning (empty) result state: ", verbose);
+        logHitGroup(result, verbose);
+
+        logIfVerbose(String.format("alpha is %.2f", alpha), verbose);
+        logIfVerbose(String.format("k is %d", k), verbose);
+
+        // Iterate through tensor hits list
+
+        int rank = 1;
+        if (alpha > 0.0) {
+            logIfVerbose(
+                    String.format(
+                            "Iterating through tensor result list. Size: %d", hitsTensor.size()),
+                    verbose);
+
+            for (Hit hit : hitsTensor) {
+                logIfVerbose(
+                        String.format("Tensor hit at rank: %d", rank),
+                        verbose); // TODO: Expose marqo__id
+                logIfVerbose(hit.toString(), verbose);
+
+                reciprocalRank = alpha * (1.0 / (rank + k));
+                rrfScores.put(
+                        hit.getId().toString(), reciprocalRank); // Store hit's score via its URI
+                hit.setField(
+                        "marqo__raw_tensor_score",
+                        hit.getRelevance()
+                                .getScore()); // Encode raw score for Marqo debugging purposes
+                hit.setRelevance(reciprocalRank); // Update score to be weighted RR (tensor)
+                result.add(hit);
+                logIfVerbose(String.format("Set relevance to: %.7f", reciprocalRank), verbose);
+                logIfVerbose("Current result state: ", verbose);
```

Comment:
It might be too costly to build this list iteratively. It might print the hit group a few hundred time.

Reply:
removed